src/Riok.Mapperly/Descriptors/Mappings/DerivedExistingTargetTypeMapping.cs

```diff
@@ -0,0 +1,83 @@
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Riok.Mapperly.Descriptors.Mappings.ExistingTarget;
+using Riok.Mapperly.Emit.Syntax;
+using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+using static Riok.Mapperly.Emit.Syntax.SyntaxFactoryHelper;
+
+namespace Riok.Mapperly.Descriptors.Mappings;
+
+/// <summary>
+/// A derived type mapping maps one base type or interface to another
+/// by implementing a if with instance checks over known types and performs the provided mapping for each type.
+/// </summary>
+public class DerivedExistingTargetTypeMapping : ExistingTargetMapping
+{
+    private const string SourceName = "source";
+    private const string TargetName = "target";
+    private const string GetTypeMethodName = nameof(GetType);
+    private readonly IReadOnlyCollection<IExistingTargetMapping> _existingTargetTypeMappings;
+
+    public DerivedExistingTargetTypeMapping(
+        ITypeSymbol sourceType,
+        ITypeSymbol targetType,
+        IReadOnlyCollection<IExistingTargetMapping> existingTargetTypeMappings
+    )
+        : base(sourceType, targetType)
+    {
+        _existingTargetTypeMappings = existingTargetTypeMappings;
+    }
+
+    public override IEnumerable<StatementSyntax> Build(TypeMappingBuildContext ctx, ExpressionSyntax target)
+    {
+        var sourceExpression = TupleExpression(CommaSeparatedList(Argument(ctx.Source), Argument(target)));
+        var sections = _existingTargetTypeMappings.Select(x => BuildSwitchSection(ctx, x)).Append(BuildDefaultSwitchSection(ctx, target));
+
+        yield return ctx.SyntaxFactory.SwitchStatement(sourceExpression, List(sections)).AddLeadingLineFeed(ctx.SyntaxFactory.Indentation);
+    }
+
+    private SwitchSectionSyntax BuildSwitchSection(TypeMappingBuildContext ctx, IExistingTargetMapping mapping)
+    {
+        var (sectionCtx, sourceVariableName) = ctx.WithNewScopedSource(SourceName);
+        var targetVariableName = sectionCtx.NameBuilder.New(TargetName);
+        sectionCtx = sectionCtx.AddIndentation();
+
+        // (A source, B target)
+        var positionalTypeMatch = PositionalPatternClause(
+            CommaSeparatedList(
+                Subpattern(DeclarationPattern(mapping.SourceType, sourceVariableName)),
+                Subpattern(DeclarationPattern(mapping.TargetType, targetVariableName))
+            )
+        );
+        var pattern = RecursivePattern().WithPositionalPatternClause(positionalTypeMatch);
+
+        // case (A source, B target):
```

Comment:
Why is the case syntax needed here instead of the switch expression like in the `DerivedTypeSwitchMapping`? Is it because #626 is not yet implemented and Mapperly does not support generating internal map methods for existing targets? Not sure if we then should implement this first.

Reply:
Switch expressions require a return type. Even if #626 were implemented it wouldn't be possible